### PR TITLE
Text from decision.

### LINKED ISSUE #2693: Clarification on the definition of choicesAllMatching
@HansOlsson, please consider this test package. 
```
package A
  model B
    package C
      extends Base;
      constant Real q = 10;
    end C;
  end B;

  partial package Base
    function f
      input Real x;
      output Real y;
    algorithm 
      y := 2*x;
    end f;
  end Base;

  partial package BaseAdvanced
    extends Base;
    constant Real p = 2;
  end BaseAdvanced;

  package Regular
    extends Base;
    constant Real w = 0;
  end Regular;

  model M
    replaceable package P = Base annotation(choicesAllMatching = true);
    Real x = P.f(time);
  end M;

  model S1
    A.M m annotation (
      Placement(visible = true, transformation(origin = {0, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  end S1;

  model S2
    A.M m(redeclare package P = A.B.C) annotation (
      Placement(visible = true, transformation(origin = {0, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  end S2;
end A;
```

According to [Section 7.3.4](https://specification.modelica.org/master/inheritance-modification-and-redeclaration.html#annotation-choices-for-suggested-redeclarations-and-modifications) of the Specification, the choices menu of `M.P` should show "_the names of all classes that are either directly or indirectly derived by inheritance from the constraining class of the declaration_".

Since the constraining class of `A.M.P` is `A.Base`, according to that definition of `choicesAllMatching` I would expect to see

1. `A.Base`
2. `A.BaseAdvanced`
3. `A.B.C`, which can be looked up according to [Section 5.3.2](https://specification.modelica.org/v3.4/Ch5.html#composite-name-lookup) of the Specification, and that in fact works just fine in `A.S2` when using Dymola 2021
4. `A.Regular`

However, if I open `A.S1` with Dymola 2021 and double-click on `m`, I only see the fourth class in the replaceable class drop-down menu, not the other three ones.

With reference to `Modelica.Media` replaceable packages, Dymola's behaviour indeed makes a lot of sense to me: in the drop-down menu I don't want to see any partial package (which won't work), nor I want to use media models that are defined inside some components of `Modelica.Fluid`, because using them is probably not such a good idea, even though it is legal according to the lookup rules.

It seems to me that the choices shown by Dymola (which I am actually fine with) correspond to this description: "_the names of all classes A.B...X.Z that are either directly or indirectly derived by inheritance from the constraining class of the declaration, **such that all classes A to X are non-partial packages and Z is non-partial**_"

Should we update Section 7.3.4 along these lines?

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #2693

### CODE REVIEW

File: chapters/inheritance.tex [hunk 1]
```diff
@@ -1060,8 +1060,10 @@ \subsection{Annotation Choices for Suggested Redeclarations and Modifications}\l
 input the values \lstinline!false! or \lstinline!true! in the graphical user interface.
 
 Choices menus of replaceable elements can be automatically constructed
-showing the names of all classes that are either directly or indirectly
-derived by inheritance from the constraining class of the declaration.
+and at least is should be possible to show the names of all classes, \lstinline!A.B.$\ldots$.X.Z!,
```
Comment (henrikt-ma) [inline chapters/inheritance.tex:1063]: Typo:
```suggestion
and at least it should be possible to show the names of all classes, \lstinline!A.B.$\ldots$.X.Z!, 
```

File: chapters/inheritance.tex [hunk 2]
```diff
@@ -1060,8 +1060,10 @@ \subsection{Annotation Choices for Suggested Redeclarations and Modifications}\l
...
 
 Choices menus of replaceable elements can be automatically constructed
-showing the names of all classes that are either directly or indirectly
-derived by inheritance from the constraining class of the declaration.
+and at least is should be possible to show the names of all classes, \lstinline!A.B.$\ldots$.X.Z!, 
+that are either directly or indirectly
+derived by inheritance from the constraining class of the declaration, where \lstinline!A! to \lstinline!X!
+are non-partial packages, and \lstinline!Z! is non-partial.
 This can be recommended by having \lstinline!annotation choicesAllMatching = true!; and disabled by having
 \lstinline!annotation choicesAllMatching = false!.
```
Comment (henrikt-ma) [inline chapters/inheritance.tex:1068]: With the discussion fresh in mind, I get the intended meaning of _be possible to show_, but I'm not sure someone who wasn't part of the discussion would understand.

Maybe it's easier to write something where the annotation is mentioned first:
> Given the annotation `choicesAllMatching = true` on a replaceable element, tools are requested to construct a menu with choices of elements usable for replacement.  Exact criteria for inclusion in such a menu are not defined, but there shall be a a way to at least get a selection of classes `A.B.….X.Z` that are either…

**Review by henrikt-ma — COMMENTED** [review]
Just leaving comments, as I have too limited experience of using this feature in actual modeling work.

File: chapters/inheritance.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/inheritance.tex:1068]: Yes, that makes a lot more sense. I will update accordingly.

**Comment by HansOlsson** [comment]
Please review @otronarp @casella and/or @gkurzbach or others that are familiar with choicesAllMatching.

**Review by eshmoylova — APPROVED** [review]

**Comment by ooer** [comment]
I would like to specify that the plug compatibility with the constraining type is the criteria. We could give the hint that all derived classes from the constraining class are plug compatible by definition.

**Comment by gkurzbach** [comment]
> I would like to specify that the plug compatibility with the constraining type is the criteria. We could give the hint that all derived classes from the constraining class are plug compatible by definition.

Thats also my comment.

**Pull request merged**